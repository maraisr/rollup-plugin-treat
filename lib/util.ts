import type { OutputChunk, RollupCache } from 'rollup';
import { rollup } from 'rollup';
import resolve from '@rollup/plugin-node-resolve';
import builtins from 'builtin-modules';

import { diary } from 'diary';

export const d = diary('rollup-plugin-treat').debug;

export const childCompile = async (id: string, cache?: RollupCache): Promise<OutputChunk> => {
	const fileBundle = await rollup({
		input: id,
		external: [...builtins, 'treat'],
		plugins: [resolve()],
		cache
	});

	const { output } = await fileBundle.generate({
		format: 'cjs',
	});

	if (output.length > 1)
		throw new Error(
			"Didnt expect child compiler to produce more than 1 file. Perhaps you're using async imports?",
		);

	d('compiled treat file %s', id);

	return output[0];
};
