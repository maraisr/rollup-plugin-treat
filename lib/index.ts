import type { FilterPattern } from '@rollup/pluginutils';
import { createFilter, dataToEsm } from '@rollup/pluginutils';
import crypto from 'crypto';
import eval from 'eval';
import path from 'path';
import { TransformResult } from 'rollup';
import type { Plugin } from 'rollup';
import { Adapter } from 'treat/dist/declarations/src/types';
import type { StoreType, StyleItem } from './store';
import { store, STYLE_TYPE } from './store';
import { childCompile, d } from './util';
import {walk} from 'astray';

const { version } = require('../package.json');

interface FilterOptions {
	include?: FilterPattern;
	exclude?: FilterPattern;
}

interface Options extends FilterOptions {
	localIdentName: string;
	themeIdentName: (() => string) | string;
	outputCSS: boolean | string;
	minify: boolean;
	browsers: Array<string>;

	onCSSOutput(source: string): Promise<string> | string;
}

const isProduction = !process.env.ROLLUP_WATCH;

const defaultOptions = (options: Partial<Options>): Options => {
	const {
		exclude = null,
		outputCSS = true,
		include = /\.treat\.(ts|js)$/,
		localIdentName = !isProduction
			? '[name]-[local]_[hash:base64:5]'
			: '[hash:base64:5]',
		themeIdentName = !isProduction
			? '_[name]-[local]_[hash:base64:4]'
			: '[hash:base64:4]',
		minify = isProduction,
		browsers = [],
		onCSSOutput = (source) => source,
	} = options;

	return {
		exclude,
		localIdentName,
		themeIdentName,
		include,
		outputCSS,
		onCSSOutput,
		minify,
		browsers,
	};
};

export const rollupPluginTreat = (passedOptions: Partial<Options>): Plugin => {
	d('version %s', version);
	d('environment %s', isProduction ? 'production' : 'development');

	const options = defaultOptions(passedOptions);
	const shouldProcessAsTreat = createFilter(options.include, options.exclude);

	const treatStore = store();

	const treats = new Map,
		parents = new Map;

	return {
		name: 'TreatPlugin',

		async resolveId( importee, importer) {
			// TODO: Use shouldProcessAsTreat here
			if (/\.treat/.test(importee)) {
				let entry = parents.get(importer) || { ref: null, order: new Set };
				let absolute = await this.resolve(importee, importer, {skipSelf: true});
				if (absolute) entry.order.add(absolute);
				parents.set(importer, entry);
				debugger;
			}

			return null;
		},

		async transform(_code: string, id) {
			if (shouldProcessAsTreat(id)) {

			}

			const entry = parents.get(id);
			debugger;
		},

		async generateBundle(_outputOptions, _bundle) {
			if (!options.outputCSS) return;

			let output = '';

			for await (const moduleId of this.getModuleIds()) {
				const maybeResource = treatStore.getStyleResource(
					moduleId
				);
				debugger;
			}
			/*const modules = this.getModuleIds().map((m) =>
				this.getModuleInfo(m),
			);*/

			debugger;

			/*for (const module of modules) {
				const maybeResource = treatStore.getStyleResource(
					module.id,
				);

				if (maybeResource) {
					const one_object = maybeResource.reduce(
						(result, item) => {
							return {
								...result,
								...item.styles,
							};
						},
						{},
					);

					const css =
						(await processCss(one_object, {
							browsers: options.browsers,
							minify: options.minify,
							from: module.id,
						})) ?? '';

					output += css;
				}
			}

			const passedCss = await options.onCSSOutput(output);

			this.emitFile({
				type: 'asset',
				source: passedCss,
				fileName:
					typeof options.outputCSS === 'string'
						? options.outputCSS
						: undefined,
			});*/
		},
	};
};

const processTreatFile = (
	sourceCode: string,
	context: { id: string; options: Options; store: StoreType },
) => {
	const trackingStyles = new Set<StyleItem>();

	const __store__:Adapter = {
		addTheme(theme) {
			context.store.addTheme(theme);
		},

		getThemes: () => context.store.getThemes(),

		addLocalCss(styles) {
			trackingStyles.add({
				type: STYLE_TYPE.LOCAL,
				styles,
			});
		},
		addThemedCss(themeRef, styles) {
			trackingStyles.add({
				type: STYLE_TYPE.THEME,
				themeRef,
				styles,
			});
		},

		getIdentName(localName, scopeId, theme) {
			return typeof theme === 'undefined'
				? processIdent(context.options.localIdentName, context.id)(
					localName,
					scopeId,
					theme,
				)
				: processIdent(context.options.themeIdentName, context.id)(
					localName,
					scopeId,
					theme,
				);
		},
	};

	let result;
	try {
		result = eval(`require('treat/adapter').setAdapter(__store__);\n${sourceCode}`,
			context.id,
			{ console, __store__ },
			true,
		);
	} catch (e) {
		throw e;
	}
	debugger;
	context.store.addStyleResource(context.id, trackingStyles);

	return result;
};

const processIdent = (identFn: ((theme: string) => string) | string, id: string) => {
	return (localName: string, scopeId: number, theme?: string): string => {
		const identName =
			typeof identFn === 'function' ? identFn(theme) : identFn;

		return identName
			.replace(
				/\[(?:([^:\]]+):)?(?:hash|contenthash)(?::([a-z]+\d*))?(?::(\d+))?\]/gi,
				(_all, hashType = 'md5', digestType, maxLength = 10) =>
					crypto
						.createHash(hashType)
						.update(`${id}${scopeId}`)
						.digest(digestType)
						.slice(0, maxLength),
			)
			.replace(/\[local]/gi, localName)
			.replace(/\[name]/gi, path.basename(id, path.extname(id)))
			.replace('treat-', '')
			.replace(new RegExp('[^a-zA-Z0-9\\-_\u00A0-\uFFFF]', 'g'), '-')
			.replace(/^((-?[0-9])|--)/, '_$1');
	};
};
