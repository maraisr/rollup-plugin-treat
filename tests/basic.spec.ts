import { rollup } from 'rollup';
import fs from 'fs';
import { resolve } from 'path';

import { rollupPluginTreat } from '../lib';

const fixtureFolder = resolve(__dirname, './fixtures/');

const fixtureInputs = fs.readdirSync(fixtureFolder, {
	withFileTypes: false,
});

const build = (options, testFolder) =>
	rollup({
		input: resolve(fixtureFolder, testFolder, './main.js'),
		external: ['treat'],
		plugins: [rollupPluginTreat(options)],
	});

// @ts-ignore
test.each(fixtureInputs as any)(
	'it should snapshot match %s',
	async (testFolder: string) => {
		const bundle = await build(
			{
				outputCSS: 'bundle.css',
			},
			testFolder,
		);

		const { output } = await bundle.generate({
			hoistTransitiveImports: true,
			format: 'esm',
		});

		output.forEach((item) => {
			if (item.type === 'chunk') {
				expect(item.code).toMatchSnapshot(item.name);
			} else if (item.type === 'asset') {
				expect(item.source).toMatchSnapshot(item.fileName);
			}
		});
	},
);

// @ts-ignore
test.each(fixtureInputs as any)(
	'it should not output css %s',
	async (testFolder: string) => {
		const onCSSOutputSpy = jest.fn();

		const bundle = await build(
			{
				outputCSS: false,
				onCSSOutput: onCSSOutputSpy,
			},
			testFolder,
		);

		const { output } = await bundle.generate({
			hoistTransitiveImports: true,
			format: 'esm',
		});

		expect(output.length).toBeGreaterThan(0);
		expect(onCSSOutputSpy).toHaveBeenCalledTimes(0);
	},
);
