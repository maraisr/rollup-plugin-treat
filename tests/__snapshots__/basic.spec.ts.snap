// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should snapshot match async-modules: a 1`] = `
"import { resolveStyles } from 'treat';

const a = \\"a-style_EB8aX\\";
var a_treat = {
	a: a
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	a: a,
	'default': a_treat
});

console.log(
	resolveStyles(null, styleRefs),
);
"
`;

exports[`it should snapshot match async-modules: bundle.css 1`] = `".a-style_EB8aX{color:red}"`;

exports[`it should snapshot match async-modules: main 1`] = `
"import('./a-397bd8b2.js')
	.then(() => {

	});
"
`;

exports[`it should snapshot match basic-no-theme: bundle.css 1`] = `".main-style_-O-a2{color:red}"`;

exports[`it should snapshot match basic-no-theme: main 1`] = `
"import { resolveStyles } from 'treat';

const myClass = \\"main-style_-O-a2\\";
var main_treat = {
	myClass: myClass
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	myClass: myClass,
	'default': main_treat
});

console.log(
	resolveStyles(null, styleRefs),
);
"
`;

exports[`it should snapshot match basic-with-theme: bundle.css 1`] = `".main-style_YnB3c{color:pink}.main-style_4mkE9_theme-theme_hdkV{color:red}"`;

exports[`it should snapshot match basic-with-theme: main 1`] = `
"import { resolveStyles } from 'treat';

const myTheme = \\"_theme-theme_hdkV\\";

const plainClass = \\"main-style_YnB3c\\";
const themedClass = \\"$main-style_4mkE9\\";
var main_treat = {
	plainClass: plainClass,
	themedClass: themedClass
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	plainClass: plainClass,
	themedClass: themedClass,
	'default': main_treat
});

console.log(
	resolveStyles(myTheme, styleRefs),
);
"
`;

exports[`it should snapshot match module-order: bundle.css 1`] = `".a-style_e7kHH{color:red}.b-style_GDtJa{color:#00f}"`;

exports[`it should snapshot match module-order: main 1`] = `
"import { resolveStyles } from 'treat';

const a = \\"a-style_e7kHH\\";
var a_treat = {
	a: a
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	a: a,
	'default': a_treat
});

const b = \\"b-style_GDtJa\\";
var b_treat = {
	b: b
};

var styleRefs$1 = /*#__PURE__*/Object.freeze({
	__proto__: null,
	b: b,
	'default': b_treat
});

console.log(
	resolveStyles(null, styleRefs$1),
);

console.log(
	resolveStyles(null, styleRefs),
);
"
`;

exports[`it should snapshot match using-node-modules: bundle.css 1`] = `".main-style_nzFr5{color:rgba(255,205,100,.7)}"`;

exports[`it should snapshot match using-node-modules: main 1`] = `
"import { resolveStyles } from 'treat';

const myClass = \\"main-style_nzFr5\\";
var main_treat = {
	myClass: myClass
};

var styleRefs = /*#__PURE__*/Object.freeze({
	__proto__: null,
	myClass: myClass,
	'default': main_treat
});

console.log(
	resolveStyles(null, styleRefs),
);
"
`;
